# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/push/PushManager.cpp
# Commit: a8c5441fd896
# Full Hash: a8c5441fd89658dea45d939b1e930cf3d148fafd
# Author: Eden Chuang <echuang@mozilla.com>
# Date: 2024-01-26 21:47:24
# Description:
#   Bug 1873573 - Rename PromiseWorkerProxy::WorkerPromise() to GetWorkerPromise(). r=dom-worker-reviewers,smaug
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D199323
# ==============================================================================

diff -r 533ecc686bac -r a8c5441fd896 dom/push/PushManager.cpp
--- a/dom/push/PushManager.cpp	Fri Jan 26 10:41:43 2024 +0000
+++ b/dom/push/PushManager.cpp	Fri Jan 26 10:48:56 2024 +0000
@@ -113,13 +113,11 @@
         mAppServerKey(std::move(aAppServerKey)) {}
 
   bool WorkerRun(JSContext* aCx, WorkerPrivate* aWorkerPrivate) override {
-    {
-      MutexAutoLock lock(mProxy->Lock());
-      if (mProxy->CleanedUp()) {
-        return true;
-      }
+    RefPtr<Promise> promise = mProxy->GetWorkerPromise();
+    // Once Worker had already started shutdown, workerPromise would be nullptr
+    if (!promise) {
+      return true;
     }
-    RefPtr<Promise> promise = mProxy->WorkerPromise();
     if (NS_SUCCEEDED(mStatus)) {
       if (mEndpoint.IsEmpty()) {
         promise->MaybeResolve(JS::NullHandleValue);
@@ -308,8 +306,10 @@
   bool WorkerRun(JSContext* aCx, WorkerPrivate* aWorkerPrivate) override {
     MOZ_ASSERT(aWorkerPrivate);
     aWorkerPrivate->AssertIsOnWorkerThread();
-
-    RefPtr<Promise> promise = mProxy->WorkerPromise();
+    RefPtr<Promise> promise = mProxy->GetWorkerPromise();
+    if (!promise) {
+      return true;
+    }
     if (NS_SUCCEEDED(mStatus)) {
       promise->MaybeResolve(mState);
     } else {